File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedBlockEvent.java
package net.easecation.ghosty.recording.level.updated;

import cn.nukkit.level.Level;
import cn.nukkit.math.BlockVector3;
import cn.nukkit.network.protocol.BlockEventPacket;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.recording.level.LevelRecordNode;

public class LevelUpdatedBlockEvent implements LevelUpdated {

    public int x;
    public int y;
    public int z;
    public int eventType;
    public int eventData;

    private LevelUpdatedBlockEvent(int evid, float x, float y, float z, int data) {
        this.x = (int) x;
        this.y = (int) y;
        this.z = (int) z;
        this.eventType = evid;
        this.eventData = data;
    }

    public static LevelUpdatedBlockEvent of(BlockEventPacket packet) {
        return new LevelUpdatedBlockEvent(packet.eventType, packet.x, packet.y, packet.z, packet.eventData);
    }

    public LevelUpdatedBlockEvent(BinaryStream stream) {
        this.read(stream);
    }

    @Override
    public int getUpdateTypeId() {
        return LevelUpdated.TYPE_BLOCK_EVENT;
    }

    @Override
    public void processTo(LevelRecordNode node) {
        BlockEventPacket packet = new BlockEventPacket();
        packet.x = this.x;
        packet.y = this.y;
        packet.z = this.z;
        packet.eventType = this.eventType;
        packet.eventData = this.eventData;
        node.handleLevelChunkPacket(Level.chunkHash(x >> 4, z >> 4), packet);
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putBlockVector3(this.x, this.y, this.z);
        stream.putVarInt(this.eventType);
        stream.putVarInt(this.eventData);
    }

    @Override
    public void read(BinaryStream stream) {
        BlockVector3 pos = stream.getBlockVector3();
        this.x = pos.x;
        this.y = pos.y;
        this.z = pos.z;
        this.eventType = stream.getVarInt();
        this.eventData = stream.getVarInt();
    }

    @Override
    public String toString() {
        return "LevelUpdatedBlockEvent{" +
            "x=" + x +
            ", y=" + y +
            ", z=" + z +
            ", eventType=" + eventType +
            ", eventData=" + eventData +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedBlockEvent.kt
package net.easecation.ghosty.recording.level.updated

import cn.nukkit.level.Level
import cn.nukkit.math.BlockVector3
import cn.nukkit.network.protocol.BlockEventPacket
import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.recording.level.LevelRecordNode

@Serializable
data class LevelUpdatedBlockEvent(
    val x: Int,
    val y: Int,
    val z: Int,
    val eventType: Int,
    val eventData: Int,
) : LevelUpdated {

    override fun getUpdateTypeId(): Int = LevelUpdated.TYPE_BLOCK_EVENT

    override fun processTo(node: LevelRecordNode) {
        BlockEventPacket().also {
            it.x = x
            it.y = y
            it.z = z
            it.eventType = eventType
            it.eventData = eventData
        }.also { packet ->
            node.handleLevelChunkPacket(Level.chunkHash(x shr 4, z shr 4), packet)
        }
    }

    override fun backwardTo(node: LevelRecordNode) = Unit

    companion object {
        @JvmStatic
        fun of(packet: BlockEventPacket) =
            LevelUpdatedBlockEvent(packet.eventType, packet.x, packet.y, packet.z, packet.eventData)

        @JvmStatic
        val ADAPTER: LevelUpdateAdapter<LevelUpdatedBlockEvent> = Adapter
    }

    private object Adapter : LevelUpdateAdapter<LevelUpdatedBlockEvent> {
        override fun write(updated: LevelUpdatedBlockEvent, stream: BinaryStream) {
            stream.putBlockVector3(updated.x, updated.y, updated.z)
            stream.putVarInt(updated.eventType)
            stream.putVarInt(updated.eventData)
        }

        override fun read(stream: BinaryStream, formatVersion: Int): LevelUpdatedBlockEvent {
            val pos = stream.blockVector3
            return LevelUpdatedBlockEvent(
                x = pos.x,
                y = pos.y,
                z = pos.z,
                eventType = stream.varInt,
                eventData = stream.varInt
            )
        }
    }
}
